applybn/feature_generation/bn_feature_generator.py

```diff
@@ -0,0 +1,158 @@
+import random
+import numpy as np
+import pandas as pd
+from pgmpy.models import BayesianNetwork
+from pgmpy.estimators import K2Score, BayesianEstimator, BicScore
+from pgmpy.inference import VariableElimination
+
+
+class BNFeatureGenerator:
+    def __init__(self, known_structure=None):
+        self.known_structure = known_structure
+        self.bn = None
+        self.variables = None
+        self.num_classes = None
+
+    def fit(self, data, target=None, black_list=None):
+        if not isinstance(data, pd.DataFrame):
+            data = pd.DataFrame(data)
+
+        self.variables = data.columns.tolist()
+
+        if target is not None:
+            if not isinstance(target, pd.Series):
+                target = pd.Series(target)
+            self.num_classes = len(target.unique())
+
+        if self.known_structure:
+            self.bn = BayesianNetwork(self.known_structure)
+            self.bn.fit(data, estimator=BayesianEstimator)
+        else:
+            constructor = self._BayesianNetworkConstructor(data, black_list)
+            self.bn = constructor.construct_network()
+
+    def generate_features(self, data):
+        if not isinstance(data, pd.DataFrame):
+            data = pd.DataFrame(data)
+
+        inference = VariableElimination(self.bn)
+        new_features = []
+
+        for _, instance in data.iterrows():
```

Comment:
Nested loops and row-by-row iterations do us no good in terms of time complexity. If they can be reduced, please, try. Here is a few things that can be used to do so:

1. Pandas .apply method (still uses loops under the hood, but better than nothing)
2. Python [`concurrent` module](https://docs.python.org/3/library/concurrent.futures.html) to iterate rows in parallel threads.